server/src/main/java/org/apache/iotdb/db/engine/migration/MigratingFileLogManager.java

```diff
@@ -0,0 +1,127 @@
+
+  private static File MIGRATING_LOG_DIR =
+      SystemFileFactory.INSTANCE.getFile(
+          Paths.get(FilePathUtils.regularizePath(config.getSystemDir()), "migration", "migrating")
+              .toString());
+
+  protected MigratingFileLogManager() {
+    if (MIGRATING_LOG_DIR == null) {
+      logger.error("MIGRATING_LOG_DIR is null");
+    }
+
+    if (!MIGRATING_LOG_DIR.exists()) {
+      if (MIGRATING_LOG_DIR.mkdirs()) {
+        logger.info("MIGRATING_LOG_DIR {} created successfully", MIGRATING_LOG_DIR);
+      } else {
+        logger.error("MIGRATING_LOG_DIR {} create error", MIGRATING_LOG_DIR);
+      }
+      return;
+    }
+
+    if (!MIGRATING_LOG_DIR.isDirectory()) {
+      logger.error("{} already exists but is not directory", MIGRATING_LOG_DIR);
+    }
+  }
+
+  // singleton
+  private static class MigratingFileLogManagerHolder {
+    private MigratingFileLogManagerHolder() {}
+
+    private static final MigratingFileLogManager INSTANCE = new MigratingFileLogManager();
+  }
+
+  public static MigratingFileLogManager getInstance() {
+    return MigratingFileLogManagerHolder.INSTANCE;
+  }
+
+  /** started migrating tsfile and its resource/mod files */
+  public boolean start(File tsfile, File targetDir) throws IOException {
+    File logFile = SystemFileFactory.INSTANCE.getFile(MIGRATING_LOG_DIR, tsfile.getName() + ".log");
+    if (!logFile.createNewFile()) {
+      return false;
+    }
+
+    FileOutputStream logFileOutput = new FileOutputStream(logFile);
+    ReadWriteIOUtils.write(tsfile.getAbsolutePath(), logFileOutput);
+    ReadWriteIOUtils.write(targetDir.getAbsolutePath(), logFileOutput);
+    logFileOutput.flush();
+    logFileOutput.close();
+
+    return true;
+  }
+
+  /** finished migrating tsfile and related files */
+  public void finish(File tsfile) {
+    File logFile = SystemFileFactory.INSTANCE.getFile(MIGRATING_LOG_DIR, tsfile.getName() + ".log");
+    if (logFile.exists()) {
+      logFile.delete();
+    }
+  }
+
+  /** finish the unfinished MigrationTasks using log files under MIGRATING_LOG_DIR */
+  public void recover() throws IOException {
+    for (File logFile : MIGRATING_LOG_DIR.listFiles()) {
+      FileInputStream logFileInput = new FileInputStream(logFile);
+
+      String tsfilePath = ReadWriteIOUtils.readString(logFileInput);
+      String targetDirPath = ReadWriteIOUtils.readString(logFileInput);
+
+      File tsfile = SystemFileFactory.INSTANCE.getFile(tsfilePath);
+      File targetDir = SystemFileFactory.INSTANCE.getFile(targetDirPath);
+
+      if (targetDir.exists()) {
+        if (!targetDir.isDirectory()) {
+          logger.error("target dir {} not a directory", targetDirPath);
+          return;
+        }
+      } else if (!targetDir.mkdirs()) {
+        logger.error("create target dir {} failed", targetDirPath);
+        return;
+      }
+
+      TsFileResource resource = new TsFileResource(tsfile);
+      resource.migrate(targetDir);
+      finish(tsfile);
+    }
+  }
```

Comment:
There problems as below:
1. It‘s not a good idea to use TsFileResource.migrate here because this method migrates .tsfile, .resource and .mods files at the same time. If any file is absent, this method will throw an eception.
2. The for statement will exit when any error occurs in the loop, you should catch the exceptions, deal with errors and then continue the loop.
3. Add some test to simulate some situations that .tsfile, .resource or .mods file is absent.

Reply:
1. Added checks for File.exist() in TsFileResource.migrate so it should be fine here.
2. Fixed.
3. Added.